File: test-kra.py
import urllib.request
import json
import base64

consumer_key = '4TKwD3C2IkHWWoO5VEoSgBda3zfbOLIGfwU6bkquNsq6d6q3'
consumer_secret = '5j03GirNnek3nA0Eel5WWpUNO779MIAR1rfzLTL9OxAcEEK5fm6QyzTrHK1i35kb'
creds = f"{consumer_key}:{consumer_secret}".encode('utf-8')
b64_creds = base64.b64encode(creds).decode('utf-8')

token_url = "https://api.kra.go.ke/v1/token/generate?grant_type=client_credentials"
token_req = urllib.request.Request(token_url, headers={'Authorization': f'Basic {b64_creds}'})

try:
    with urllib.request.urlopen(token_req) as response:
        token = json.loads(response.read().decode())['access_token']
except Exception as e:
    exit(1)

endpoint = "https://api.kra.go.ke/checker/v1/pinbypin"
payloads = [
    {"KRAPIN": "A002345678B"},
    {"PIN": "A002345678B"},
    {"pin": "A002345678B"},
    {"TaxpayerPIN": "A002345678B"}
]

for p in payloads:
    data = json.dumps(p).encode('utf-8')
    req = urllib.request.Request(endpoint, data=data, headers={
        'Authorization': f'Bearer {token}',
        'Content-Type': 'application/json'
    })
    try:
        with urllib.request.urlopen(req) as response:
            print(f"Success for {p}:", response.status, response.read().decode())
    except urllib.error.HTTPError as e:
        print(f"HTTP Error {e.code} for {p}: {e.read().decode()}")
    except Exception as e:
        print(f"Error for {p}: {e}")

File: test-kra.js
const https = require('https');
require('dotenv').config({ path: '.env' });

function request(url, options, body) {
  return new Promise((resolve, reject) => {
    const req = https.request(url, options, res => {
      let data = '';
      res.on('data', chunk => data += chunk);
      res.on('end', () => resolve({ status: res.statusCode, data }));
    });
    req.on('error', reject);
    if (body) req.write(body);
    req.end();
  });
}

async function test() {
    try {
        const pinType = 'pinByPIN';
        const consumerKey = process.env.KRA_PIN_CONSUMER_KEY;
        const consumerSecret = process.env.KRA_PIN_CONSUMER_SECRET;
        const tokenEndpoint = "https://api.kra.go.ke/v1/token/generate?grant_type=client_credentials";

        console.log("Fetching token...");
        const credentials = Buffer.from(`${consumerKey}:${consumerSecret}`).toString('base64');
        const tokenRes = await request(tokenEndpoint, {
            method: 'GET',
            headers: {
                'Authorization': `Basic ${credentials}`,
            }
        });
        const tokenData = JSON.parse(tokenRes.data);
        console.log("Token response:", tokenData);

        const token = tokenData.access_token;
        const endpoint = `https://api.kra.go.ke/checker/v1/pinbypin`;
        const pin = 'A002345678B';
        
        console.log(`Calling ${endpoint} with KRAPIN=${pin}`);
        const body = JSON.stringify({ KRAPIN: pin });
        const res = await request(endpoint, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                'Authorization': `Bearer ${token}`,
                'User-Agent': 'Mozilla/5.0'
            }
        }, body);
        
        console.log("Response:", res.status, res.data);
    } catch (e) {
        console.error(e);
    }
}
test();
